### DISCUSSION TITLE
ShiftedBetaGeoModelIndividual - question about input data and cohorts

### DISCUSSION THREAD

**Discussion by ezietsman** [pymc-labs/pymc-marketing#681] [Q&A] [answered]
Hi, I've been trying to model customer subscription data using the ShiftedBetaGeoModelIndividual model. However I'm having a hard time getting the model to converge reliably and I'm wondering if it is an input data problem: The documentation says the data input should have:

    customer_id: Customer labels. There should be one unique label for each customer
    t_churn: Time at which the customer cancelled the contract (starting at 0). It should equal T for users that have not cancelled by the end of the observation period
    T: Maximum observed time period (starting at 0)
    
Lets assume I'm looking over a 12 month period, each month during that time I have a number of customers starting a subscription and a number ending their subscriptions. Am I correct in making the input data like this:

| Customer id | t_churn | T |  Notes |
|  -------          |  ----      |  --- | --- |
| 0                      |  3          |   3 |   started month 9, still active at month 12  | 
| 1                      |  3           |   12   |  cancelled after 3 months  | 
| 2                      |  12           |   12   |  Started first month, still active |

Every example I see has the value T the same for the entire cohort which tells me one is looking at users who all started at the same time, which is not the case for me. I'm trying to, eventually, calculate customer lifetime value (I also have different kinds of users but lets assume I'm looking at only one specific subset). Is this data I'm using correct for my usecase?

Any help will be greatly appreciated.

**Comment by ColtAllen** [comment, marked as answer]
Hey @ezietsman,

Seems the documentation may require clarification, so thanks for bringing this to our attention.

`T` is simply the maximum number of observed contract renewals in the dataset, and `t_churn` is independent of start date. The input data should look like this:

Customer id | t_churn | T | Notes
-- | -- | -- | --
0 | 12 | 12 | started month 9, still active at month 12
1 | 3 | 12 | cancelled after 3 months
2 | 12 | 12 | started first month, still active

If this seems confusing, it's because active customers are treated as "right-censored" (i.e., their contractual termination date is unobservable). **The time durations of inactive customers are crucial for fitting this model**. If this seems confusing at first, it'll make more sense after the predictive methods are ran:

- `model.distribution_new_customer_theta` (the dropout probability distribution for new customers)
- `model.distribution_customer_churn_time` (the probability a given customer is still active _x_ contract periods into the future)

**Reply by ezietsman** [reply]
Thank you so much. This answers my question about the data. 

Is there some documentation I can look at that will show me how to incorporate different kinds of customers into a single model, like a hierarchical model e.g. I have another data field that specifies contains the amount they pay on their contract (some people might have been given a deal when they signed up say)

**Reply by ColtAllen** [reply]
There is an issue open for this: https://github.com/pymc-labs/pymc-marketing/issues/169

Which in turn links to another issue referencing the original research paper: https://github.com/pymc-labs/pymc-marketing/issues/167

I've also found a research note which may enable adding time-varying covariates as a modeling option.

I can look into escalating this, but in the meantime if you find yourself building out a solution in `pymc` directly, please feel free to share and/or open a PR and we'll get it added to the library.

**Reply by ColtAllen** [reply]
Found another paper on hierarchical cohorts:

https://brucehardie.com/notes/017/sBG_estimation.pdf
